cabby/model/text/train.py

```diff
@@ -0,0 +1,396 @@
+
+"""Simple text classification example.
+
+This script trains a text classifier on a dataset containing paragraphs that
+are linked to either Manhattan or Pittburgh. This is just to get something
+basic that works for the geo-oriented text problems we are addressing with
+the Rendezvous task.
+
+A sentence is encoded with an LSTM and a binary prediction is made from
+the final LSTM state using an MLP.
+
+The data is loaded from raw text files and converted into TF Records.
+
+Example command line call:
+
+$ bazel-bin/cabby/model/text/train \
+  --data_dir ~/data/morp/morp-balanced  \
+  --model_dir ~/tmp/morp-model
+
+This code is copied with very few modifications from the Flax SST2 example:
+
+https://github.com/google/flax/tree/master/examples/sst2
+"""
+# pylint: disable=import-error,too-many-locals,too-many-arguments
+
+import collections
+from typing import Any, Dict, Text, Tuple
+
+from absl import app
+from absl import flags
+from absl import logging
+
+import flax
+import flax.training.checkpoints
+from flax import nn
+
+import jax
+import jax.numpy as jnp
+import numpy as np
+
+import tensorflow.compat.v2 as tf
+import tensorflow_datasets as tfds
+from tensorflow.io import gfile
+
+import input_pipeline
+import model as sst2_model
+
+
+FLAGS = flags.FLAGS
+
+flags.DEFINE_float(
+    'learning_rate', default=0.0005,
+    help=('The learning rate for the Adam optimizer.'))
+
+flags.DEFINE_integer(
+    'batch_size', default=64,
+    help=('Batch size for training.'))
+
+flags.DEFINE_integer(
+    'num_epochs', default=20,
+    help=('Number of training epochs.'))
+
+flags.DEFINE_float(
+    'dropout', default=0.5,
+    help=('Dropout rate'))
+
+flags.DEFINE_float(
+    'emb_dropout', default=0.5,
+    help=('Embedding dropout rate'))
+
+flags.DEFINE_float(
+    'word_dropout_rate', default=0.1,
+    help=('Word dropout rate. Replaces input words with <unk>.'))
+
+flags.DEFINE_string(
+    'model_dir', default=None,
+    help=('Directory to store model data'))
+
+flags.DEFINE_string(
+    'data_dir', default=None,
+    help=('Directory to read MORP data from.'))
+
+flags.DEFINE_integer(
+    'hidden_size', default=256,
+    help=('Hidden size for the LSTM and MLP.'))
+
+flags.DEFINE_integer(
+    'embedding_size', default=256,
+    help=('Size of the word embeddings.'))
+
+flags.DEFINE_integer(
+    'max_seq_len', default=55,
+    help=('Maximum sequence length in the dataset.'))
+
+flags.DEFINE_integer(
+    'min_freq', default=5,
+    help=('Minimum frequency for training set words to be in the vocabulary.'))
+
+flags.DEFINE_float(
+    'l2_reg', default=1e-6,
+    help=('L2 regularization weight'))
+
+flags.DEFINE_integer(
+    'seed', default=0,
+    help=('Random seed for network initialization.'))
+
+flags.DEFINE_integer(
+    'checkpoints_to_keep', default=1,
+    help=('How many checkpoints to keep. Default: 1 (keep best model only)'))
+
+
+@jax.vmap
+def binary_cross_entropy_loss(logit: jnp.ndarray, label: jnp.ndarray):
```

Comment:
not a review comment -- but jax does not have a default impl for xent loss? :(

Reply:
Jax certainly doesn't as it is more "bare metal" that neural network libraries build abstraction upon. Flax appears not to have default implementations for loss functions, but Trax does: 

https://trax-ml.readthedocs.io/en/latest/trax.layers.html#module-trax.layers.metrics

Not sure if they are planning to add loss functions to Flax, but do know they are working on the new Linen API for Flax: https://github.com/google/flax/tree/master/flax/linen